src/pcrlock/pcrlock.c

```diff
@@ -4164,6 +4170,127 @@ static int remove_policy_file(const char *path) {
+                        /* ret_uuid= */ NULL,
+                        /* ret_devid= */ NULL);
+        if (r < 0) {
+                if (r != -ENOKEY)
+                        return log_error_errno(r, "Failed to find XBOOTLDR partition: %m");
+
+                r = find_esp_and_warn(
+                                /* root= */ NULL,
+                                /* path= */ NULL,
+                                /* unprivileged_mode= */ false,
+                                &path,
+                                /* ret_part= */ NULL,
+                                /* ret_pstart= */ NULL,
+                                /* ret_psize= */ NULL,
+                                /* ret_uuid= */ NULL,
+                                /* ret_devid= */ NULL);
+                if (r < 0) {
+                        if (r != -ENOKEY)
+                                return log_error_errno(r, "Failed to find ESP partition: %m");
+
+                        *ret = NULL;
+                        return 0; /* not found! */
+                }
+        }
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read machine ID: %m");
+
+        r = boot_entry_token_ensure(
+                        /* root= */ NULL,
+                        "/etc/kernel",
+                        machine_id,
+                        /* machine_id_is_random = */ false,
+                        &arg_entry_token_type,
+                        &arg_entry_token);
+        if (r < 0)
+                return r;
+
+        fn = strjoin("pcrlock.", arg_entry_token, ".cred");
+        if (!fn)
+                return log_oom();
+
+        if (!filename_is_valid(fn))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Credential name '%s' would not be a valid file name, refusing.", fn);
+
+        joined = path_join(path, "loader/credentials", fn);
+        if (!joined)
+                return log_oom();
+
+        *ret = TAKE_PTR(joined);
+        return 1; /* found! */
+}
+
+static int write_boot_policy_file(const char *json_text) {
+        _cleanup_free_ char *boot_policy_file = NULL;
+        int r;
+
+        assert(json_text);
+
+        r = determine_boot_policy_file(&boot_policy_file);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                log_info("Did not find XBOOTLDR/ESP partition, not writing boot policy file.");
+                return 0;
+        }
+
+        _cleanup_free_ char *c = NULL;
+        r = path_extract_filename(boot_policy_file, &c);
+        if (r < 0)
+                return log_error_errno(r, "Failed to extract file name from %s: %m", boot_policy_file);
+
+        ascii_strlower(c); /* lowercase this file, no matter what, since stored on VFAT, and we don't want to
+                            * run into case change incompatibilities */
+
+        _cleanup_(iovec_done) struct iovec encoded = {};
+        r = encrypt_credential_and_warn(
```

Comment:
Given we know it's always going to be unencrypted, why go through the trouble of creating an encrypted credential instead of just a normal plain text one?

Reply:
that's a very valid question. 

So I didn't really want to add a concept for regularly storing plaintext creds in the ESP. Something like that shouldn't become a "thing" that people just do, if you follow what I mean. If we'd fully support this then it would become too easy to do that, and I think we should not make that easy. So my way out is to just do this via an encrypted credential, but then set the encryption to null mode. We support a null mode for encrypted creds forever, for compat with tpm-less systems. All I did is allow enabling this mode if explicitly requested also on tpm-equipped systems, for selected creds.

I mean, this special data blob on one hand should be treated like all other creds I think, i.e. stored in the ESP, picked up by sd-stub, copied into the initrd, be considered untrusted. Except that in some facet it shouldn't be treated like everything else: it should be accessible without any secret key material, since it describes the path to the secret key material. But the part where it is different seemed much smaller to me then where it should be treated like everything else.

also, esp means vfat, and vfat is a shitty fs. what encrypted creds with a null key give you over regular files is a checksumming logic if you so will: we use AEAD here, and hence we'll check the integrity of the file still, and thus IO/accidental corruptions will neatly be detected.

and then, creds carry some metadata which is nice to use here, such as validity time rules and naming checks. while these are not essential, they are nice to have.

Reply:
by making this a cred we also don't need to special case this in sd-boot btw which is nice.

Reply:
Checksumming is a nice property, yes. But I think we should have an opt-in mechanism for uncencrypted untrusted credentials, as that is going to be useful in a number of cases, like synthesizing from the boot menu and so on

Reply:
i am not convinced of that usecase, as dicussed elsewhere. untrusted creds must be an absolute exception.

Reply:
(but anyway, that's a discussion orthogonal to this one)